pkg/controller/batchrelease/control/canarystyle/interface.go

```diff
@@ -0,0 +1,45 @@
+package canarystyle
+
+import (
+	"github.com/openkruise/rollouts/api/v1alpha1"
+	batchcontext "github.com/openkruise/rollouts/pkg/controller/batchrelease/context"
+	"github.com/openkruise/rollouts/pkg/util"
+)
+
+type Interface interface {
+	CanaryInterface
+	StableInterface
+	// BuildStableController will get stable workload object and parse
+	// stable workload info, and return a controller for stable workload.
+	BuildStableController() (StableInterface, error)
+	// BuildCanaryController will get canary workload object and parse
+	// canary workload info, and return a controller for canary workload.
+	BuildCanaryController(release *v1alpha1.BatchRelease) (CanaryInterface, error)
+	// CalculateBatchContext calculate the current batch context according to
+	// our release plan and the statues of stable workload and canary workload.
+	CalculateBatchContext(release *v1alpha1.BatchRelease) *batchcontext.BatchContext
+}
+
+// CanaryInterface contains the methods about canary workload
+type CanaryInterface interface {
+	// GetCanaryInfo return the information about canary workload
+	GetCanaryInfo() *util.WorkloadInfo
+	// UpgradeBatch upgrade canary workload according to current batch context
+	UpgradeBatch(*batchcontext.BatchContext) (bool, error)
+	// CreateCanaryWorkload create canary workload before rolling out
+	CreateCanaryWorkload(controller *v1alpha1.BatchRelease) (bool, error)
+	// DeleteCanaryWorkload delete canary workload after rolling out
+	DeleteCanaryWorkload(controller *v1alpha1.BatchRelease) (bool, error)
+}
+
+// StableInterface contains the methods about stable workload
+type StableInterface interface {
+	// IsPromoted return true means all pods are upgraded, and available pods satisfy MaxUnavailable
+	IsPromoted() bool
+	// GetStableInfo return the information about stable workload
+	GetStableInfo() *util.WorkloadInfo
+	// ClaimStableWorkload claim the stable workload is under our control
```

Comment:
// ClaimStableWorkload claim the stable workload is under rollout control

// FreeStableWorkload free the stable workload from rollout control